database/engine/metadata_log/logfile.c

```diff
@@ -273,18 +273,166 @@ static int check_metadata_logfile_superblock(uv_file file)
     assert(req.result >= 0);
     uv_fs_req_cleanup(&req);
 
-    if (strncmp(superblock->magic_number, RRDENG_METALOG_MAGIC, RRDENG_MAGIC_SZ) ||
-        strncmp(superblock->version, RRDENG_METALOG_VER, RRDENG_VER_SZ)) {
+    if (strncmp(superblock->magic_number, RRDENG_METALOG_MAGIC, RRDENG_MAGIC_SZ)) {
         error("File has invalid superblock.");
         ret = UV_EINVAL;
     } else {
         ret = 0;
     }
-    error:
+    if (strncmp(superblock->version, RRDENG_METALOG_VER, RRDENG_VER_SZ)) {
+        error("File has unknown version \"%.*s\". Compatibility is not guaranteed.", RRDENG_VER_SZ,
+              superblock->version);
+    }
+error:
     free(superblock);
     return ret;
 }
 
+void replay_record(struct metadata_logfile *metalogfile, struct rrdeng_metalog_record_header *header, void *payload)
+{
+    struct metalog_instance *ctx = metalogfile->ctx;
+    char *line, *nextline, *record_end;
+    int ret;
+    struct metalog_pluginsd_state *state = ctx->metalog_parser_object->private;
+
+    info("RECORD contents: %.*s", (int)header->payload_length, (char *)payload);
+    record_end = (char *)payload + header->payload_length - 1;
+    *record_end = '\0';
+
+    for (line = payload ; line ; line = nextline) {
+        nextline = strchr(line, '\n');
+        if (nextline) {
+            *nextline++ = '\0';
+        }
+        ret = parser_action(ctx->metalog_parser_object->parser, line);
+        info("parser_action ret:%d", ret);
+        if (ret)
+            return; /* skip record due to error */
+    };
+}
+
+/* This function only works with buffered I/O */
+static inline int metalogfile_read(struct metadata_logfile *metalogfile, void *buf, size_t len, uint64_t offset)
+{
+    struct metalog_instance *ctx;
+    uv_file file;
+    uv_buf_t iov;
+    uv_fs_t req;
+    int ret;
+
+    ctx = metalogfile->ctx;
+    file = metalogfile->file;
+    iov = uv_buf_init(buf, len);
+    ret = uv_fs_read(NULL, &req, file, &iov, 1, offset, NULL);
+    if (unlikely(ret < 0 && ret != req.result)) {
```

Comment:
Why are we testing for this?

Reply:
This may be an I/O error, permission error etc. Why shouldn't we test for this?

Reply:
Ah, misread the code. This is indeed something else. If libuv has an internal error, such as out of memory, we do `fatal()`, otherwise we grab the I/O error, permission error etc and handle it.

Reply:
The `ret<0` case is worth testing for (and handled in the next block), but the `ret>=req.result` seems a bit extreme (given the `POST` macro in the uv code).